crates/ra_ide/src/completion/patterns.rs

```diff
@@ -0,0 +1,118 @@
+use ra_syntax::{
+    algo::non_trivia_sibling,
+    ast::{self, LoopBodyOwner},
+    match_ast, AstNode, Direction, NodeOrToken, SyntaxElement,
+    SyntaxKind::*,
+    SyntaxNode,
+};
+
+pub(crate) fn inside_impl(element: SyntaxElement) -> bool {
+    let node = match element {
+        NodeOrToken::Node(node) => node,
+        NodeOrToken::Token(token) => token.parent(),
+    };
+    node.ancestors().find(|it| it.kind() == IMPL_DEF).is_some()
+}
```

Comment:
We might or migth not want to add a couple of really simple smoke tests immediateely after each pattern fn, just to make it more obvious how it works.